File: src/falcon_core/physics/config/geometries/right_reservoir_with_implanted_ohmic.py
from __future__ import annotations
from typing import Any, List, Dict, Tuple, Optional
from falcon_core._capi.right_reservoir_with_implanted_ohmic import RightReservoirWithImplantedOhmic as _CRightReservoirWithImplantedOhmic
from falcon_core.physics.device_structures.connection import Connection

class RightReservoirWithImplantedOhmic:
    """Python wrapper for RightReservoirWithImplantedOhmic."""

    def __init__(self, c_obj):
        self._c = c_obj

    @classmethod
    def _from_capi(cls, c_obj):
        if c_obj is None:
            return None
        return cls(c_obj)

    @classmethod
    def new(cls, name: str, left_neighbor: Connection, ohmic: Connection) -> RightReservoirWithImplantedOhmic:
        return cls(_CRightReservoirWithImplantedOhmic.new(name, left_neighbor._c, ohmic._c))

    @classmethod
    def from_json(cls, json: str) -> RightReservoirWithImplantedOhmic:
        return cls(_CRightReservoirWithImplantedOhmic.from_json(json))

    def name(self, ) -> str:
        ret = self._c.name()
        return ret

    def type(self, ) -> str:
        ret = self._c.type()
        return ret

    def ohmic(self, ) -> Connection:
        ret = self._c.ohmic()
        if ret is None: return None
        return Connection._from_capi(ret)

    def left_neighbor(self, ) -> Connection:
        ret = self._c.left_neighbor()
        if ret is None: return None
        return Connection._from_capi(ret)

    def equal(self, b: RightReservoirWithImplantedOhmic) -> None:
        ret = self._c.equal(b._c)
        return ret

    def not_equal(self, b: RightReservoirWithImplantedOhmic) -> None:
        ret = self._c.not_equal(b._c)
        return ret

    def __eq__(self, other):
        """Operator overload for =="""
        if not isinstance(other, RightReservoirWithImplantedOhmic):
            return NotImplemented
        return self.equal(other)

    def __ne__(self, other):
        """Operator overload for !="""
        if not isinstance(other, RightReservoirWithImplantedOhmic):
            return NotImplemented
        return self.not_equal(other)

File: ocaml/_opam/.opam-switch/sources/falcon_core/src/falcon_core/physics/config/geometries/rightreservoirwithimplantedohmic.ml
open Ctypes
open Capi_bindings
open ErrorHandling

open Falcon_core.Physics.Device_structures

class c_rightreservoirwithimplantedohmic (h : unit ptr) = object(self)
  val raw_val = h
  method raw = raw_val
  initializer Gc.finalise (fun _ ->
    Capi_bindings.rightreservoirwithimplantedohmic_destroy raw_val;
    ErrorHandling.raise_if_error ()
  ) self
end

module RightReservoirWithImplantedOhmic = struct
  type t = c_rightreservoirwithimplantedohmic

  let copy (handle : t) : t =
    ErrorHandling.read handle (fun () ->
      let ptr = Capi_bindings.rightreservoirwithimplantedohmic_copy handle#raw in
      ErrorHandling.raise_if_error ();
      new c_rightreservoirwithimplantedohmic ptr
    )

  let fromjson (json : string) : t =
    let ptr = Capi_bindings.rightreservoirwithimplantedohmic_from_json_string (Capi_bindings.string_wrap json) in
    ErrorHandling.raise_if_error ();
    new c_rightreservoirwithimplantedohmic ptr

  let make (name : string) (left_neighbor : Connection.t) (ohmic : Connection.t) : t =
    ErrorHandling.multi_read [left_neighbor; ohmic] (fun () ->
      let ptr = Capi_bindings.rightreservoirwithimplantedohmic_create (Capi_bindings.string_wrap name) left_neighbor#raw ohmic#raw in
      ErrorHandling.raise_if_error ();
      new c_rightreservoirwithimplantedohmic ptr
    )

  let equal (handle : t) (other : t) : bool =
    ErrorHandling.multi_read [handle; other] (fun () ->
      let result = Capi_bindings.rightreservoirwithimplantedohmic_equal handle#raw other#raw in
      ErrorHandling.raise_if_error ();
      result
    )

  let notEqual (handle : t) (other : t) : bool =
    ErrorHandling.multi_read [handle; other] (fun () ->
      let result = Capi_bindings.rightreservoirwithimplantedohmic_not_equal handle#raw other#raw in
      ErrorHandling.raise_if_error ();
      result
    )

  let toJsonString (handle : t) : string =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.rightreservoirwithimplantedohmic_to_json_string handle#raw in
      ErrorHandling.raise_if_error ();
      Capi_bindings.string_to_ocaml result
    )

  let name (handle : t) : string =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.rightreservoirwithimplantedohmic_name handle#raw in
      ErrorHandling.raise_if_error ();
      Capi_bindings.string_to_ocaml result
    )

  let type_ (handle : t) : string =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.rightreservoirwithimplantedohmic_type handle#raw in
      ErrorHandling.raise_if_error ();
      Capi_bindings.string_to_ocaml result
    )

  let ohmic (handle : t) : Connection.t =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.rightreservoirwithimplantedohmic_ohmic handle#raw in
      ErrorHandling.raise_if_error ();
      new c_connection result
    )

  let leftNeighbor (handle : t) : Connection.t =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.rightreservoirwithimplantedohmic_left_neighbor handle#raw in
      ErrorHandling.raise_if_error ();
      new c_connection result
    )

end